### PR TITLE
make RemainingCandidates::next peekable.

### PR DESCRIPTION
**Pull request by Eh2406**
`candidates.next` always came with a `candidates.clone().next(prev_active).is_ok` so let's just make that part of `next`. no `clone` needed.

### CODE REVIEW

**Comment by Eh2406** [comment]
Thinking about it some more, an alternative is to just skip the `has_another` code and always push a cloned `BacktrackFrame` to the `backtrack_stack` then use the [`loop/if let`](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/resolver/mod.rs#L777) in `find_candidate` to find one with another candidate. Currently that `if let` is always true, as we only add an item to the `backtrack_stack` if it has another candidate.

In the happy path this is a small performance hit, a clone for every time we only have 1 candidate. But how often does that happen?

**Comment by Eh2406** [comment]
I started work on that alternative at:
https://github.com/Eh2406/cargo/tree/remove_has_another
Just to compare which leads to more clean up.

> how often does that happen?
Actually every time we are forced by a previous selection.

File: src/cargo/core/resolver/mod.rs [hunk 1]
```diff
@@ -559,15 +570,22 @@ impl RemainingCandidates {
...
+                .find(|a| compatible(a.version(), b.summary.version()))
+            {
                 if *a != b.summary {
                     self.conflicting_prev_active.insert(a.package_id().clone());
-                    continue
+                    continue;
                 }
             }
-            return Ok(b);
+            if let Some(r) = ::std::mem::replace(&mut self.has_another, Some(b)) {
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: nit: could `std::mem` be imported above?

File: src/cargo/core/resolver/mod.rs [hunk 2]
```diff
@@ -759,13 +767,7 @@ fn find_candidate<'a>(
...
 ) -> Option<Candidate> {
     while let Some(mut frame) = backtrack_stack.pop() {
-        let (next, has_another) = {
-            let prev_active = frame.context_backup.prev_active(&frame.dep);
-            (
-                frame.remaining_candidates.next(prev_active),
-                frame.remaining_candidates.clone().next(prev_active).is_ok(),
-            )
-        };
+        let next= frame.remaining_candidates.next(frame.context_backup.prev_active(&frame.dep));
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: nit: spacing on the `=` here

**Comment by Eh2406** [comment]
@alexcrichton, I fixed you nits. Then I noted that some of my cleanups from the `remove_has_another` bach can still be made. Namly, with a small change we can have only 1 place we push to `backtrack_stack`. IMO this cleans up control flow. I am working on fixing the merge conflicts.

**Comment by Eh2406** [comment]
merge done.

**Comment by alexcrichton** [comment]
Looking at some profiles from master Cargo, I think this commit unfortunately makes us lose out on a previously important optimization which is when `has_another` was false don't actually clone the `Context`. That's the fast path when using lock files (and the second round of resolution currently in Cargo), but @Eh2406 do you think we might be able to recover that?

**Comment by Eh2406** [comment]
That is not good. This PR was not intended to change that optimization. I will help track down that regression any way I can. I don't see how this commit could be causing a problem with that, the only `Context::clone` is line 754 witch is guarded by line 751. And anyway, things in the lock file are `activate`ed way back at line 647.

So I can dig in (when I next have time), what is the exact command you are testing with?
What behavior did you see before? How did you observe it? 
What behavior are you seeing now?

With that I can bisect to Identify the problem commit. I suspect it is from #5000 or #5104 as they affected clones of context. In any case, there should be ways to make it all work and have that optimization. Just need to figure out how.

**Comment by alexcrichton** [comment]
Er sorry, I think I misdiagnosed! I do indeed believe this is https://github.com/rust-lang/cargo/pull/5000 as it moved the creation of `backtrack` outside of the conditional (unconditionally cloning before `activate`).

As for a test case I was running `perf record ./x.py build` which for bad reasons runs `cargo build` a bunch of times (~10) which are all noops but work over a pretty big workspace. I found that resolution was taking a pretty big chunk of the Cargo time and I was just poking around reading code and realized that the optimization wasn't there by accident any more! 

On master though it's this line --

https://github.com/rust-lang/cargo/blob/ae796ad84a742ea39a2cb4a0c8adb0def8cee725/src/cargo/core/resolver/mod.rs#L897

which happens unconditionally before `activate` rather than inside the `has_another` conditional. Unfortunately this didn't look easy to rejigger :(

**Comment by Eh2406** [comment]
Ya, that makes sense, and will be trickier. I will look into it as soon as work allows.

**Comment by alexcrichton** [comment]
Ok awesome, thanks!

**Comment by alexcrichton** [comment]
I've opened a tracking issue for this at https://github.com/rust-lang/cargo/issues/5130

**Pull request merged**